src/test/SampleCustomLoader/sampleCustLoader.cpp

```diff
@@ -0,0 +1,324 @@
+        string xmlPath = fullpath + "/" + xmlName;
+        *xmlFile       = new char[xmlPath.length() + 1];
+        if (xmlFile == NULL) {
+            cout << "Error: could not allocate memory " << endl;
+            return CUSTLOADER_ERROR;
+        }
+        snprintf(*xmlFile, size, "%s", xmlPath.c_str());
+        cout << "xmlFile:" << *xmlFile << endl;
+    }
+    return ret;
+}
+
+int custSampleLoader::loadModel(const char* modelName, const char* basePath, const int version,
+                                   const char* loaderOptions, char** xmlBuffer, int* xmlLen,
+                                   char** binBuffer, int* binLen) {
+    cout << "custSampleLoader: Custom loadModel" << endl;
+
+    char* type    = NULL;
+    char* binFile = NULL;
+    char* xmlFile = NULL;
+
+    int ret =
+        extract_input_params((char*)basePath, version, (char*)loaderOptions, &binFile, &xmlFile);
+    if (ret != CUSTLOADER_OK || binFile == NULL || binFile == NULL) {
+        cout << "custSampleLoader: Invalid custom loader options" << endl;
+        return CUSTLOADER_ERROR;
+    }
+
+    // load models
+    ret = load_files(binFile, xmlFile, xmlBuffer, binBuffer, xmlLen, binLen);
+    if (ret != CUSTLOADER_OK || xmlBuffer == NULL || xmlBuffer == NULL) {
+        cout << "custSampleLoader: Could not read model files" << endl;
+        return CUSTLOADER_ERROR;
+    }
+
+    /* Start the watcher thread after first moel load */
+    if (watcherStarted == false){
+    	int interval = 30;
+    	startWatcher(interval);
+    	this_thread::sleep_for(chrono::seconds(1));
+    }
+    
+    models_loaded.push_back(make_pair(string(modelName),version));
+    return CUSTLOADER_OK;
+}
+
+int custSampleLoader::unloadModel(const char* modelName, int version) {
+    cout << "custSampleLoader: Custom unloadModel" << endl;
+
+    model_id_t toFind = make_pair(string(modelName),version);
+
+    auto it = models_loaded.begin();
+    for (; it != models_loaded.end(); it++){
+	    if (*it == toFind)
+		    break;
+    }
+
+    if (it == models_loaded.end()){
+	    cout << modelName << " is not loaded" <<endl;
+    }
+    else {
+	    models_loaded.erase(it);
+    }
+    return CUSTLOADER_OK;
+}
+
+void custSampleLoader::loaderDeInit() {
+    cout << "custSampleLoader: Custom loaderDeInit" << endl;
+    watcherJoin();
+    return;
+}
+
+bool custSampleLoader::getModelBlacklistStatus(const char* modelName, int version) {
+    cout << "custSampleLoader: Custom getModelBlacklistStatus" << endl;
+
+    if (models_blacklist.size() == 0)
+	    return false;
+
+    model_id_t toFind = make_pair(string(modelName), version);
+
+    auto it = models_blacklist.begin();
+    for (; it != models_blacklist.end(); it++){
+	    if (*it == toFind)
+		    break;
+    }
+
+    if (it == models_blacklist.end()){
+	return false;
+    }
+
+    /* model name and version in blacklist.. return true */
+    return true;
+}
+
+void custSampleLoader::threadFunction(future<void> futureObj) {

```

Comment:
Is my understanding correct?:
We spawn thread that checks with preset interval if model can be loaded.
Depending on answer we add such model to excluded list.
Then when ModelManager request model to be loaded/reloaded custom loader checks if it has required "access right"?

I am wondering if:
the check could happen in ModelManager thread when it tries to reload?

Reply:
Yes. The understanding is correct. When custom loader is not present, the base class will always return false for the blacklist check. I am not sure what will be the performance impact on the main thread. My understanding is it will not be big.